src/Experimental/ShortestPaths/ShortestPaths.jl

```diff
@@ -0,0 +1,105 @@
+
+using DataStructures:PriorityQueue, enqueue!, dequeue!
+
+
+struct LGEnvironment
+    threaded::Bool
+    parallel::Bool
+    LGEnvironment() = new(false, false)
+end
+     
+abstract type AbstractGraphResults end
+abstract type AbstractGraphAlgorithm end
+
+abstract type ShortestPathResults <: AbstractGraphResults end
+abstract type ShortestPathAlgorithm <: AbstractGraphAlgorithm end
+
+include("astar.jl")
+include("bellman-ford.jl")
+include("bfs.jl")
+include("desopo-pape.jl")
+include("dijkstra.jl")
+include("floyd-warshall.jl")
+include("johnson.jl")
+include("spfa.jl")
+
+
+##################
+##   Dijkstra   ##
+##################
+
+
+################################
+# Shortest Paths via algorithm #
+################################
+# if we don't pass in distances but specify an algorithm, use weights.
+shortest_paths(g::AbstractGraph, s, alg::ShortestPathAlgorithm) =
+    shortest_paths(g, s, weights(g), alg)
+
+# If we don't specify an algorithm AND there are no dists, use BFS.
+shortest_paths(g::AbstractGraph{T}, s::Integer) where {T<:Integer} = shortest_paths(g, s, BFS())
+shortest_paths(g::AbstractGraph{T}, ss::AbstractVector) where {T<:Integer} = shortest_paths(g, ss, BFS())
+
+# Full-formed methods.
+"""
+    paths(state[, vs])
+
+Given a path state `state` of type `ShortestPathResults`, return a
+vector (indexed by vertex) of the paths between the source vertex used to
+compute the path state and a single destination vertex, a list of destination
+vertices, or the entire graph. For multiple destination vertices, each
+path is represented by a vector of vertices on the path between the source and
+the destination. Nonexistent paths will be indicated by an empty vector. For
+single destinations, the path is represented by a single vector of vertices,
+and will be length 0 if the path does not exist.
+
+### Implementation Notes
+For Floyd-Warshall path states, please note that the output is a bit different,
+since this algorithm calculates all shortest paths for all pairs of vertices:
+`enumerate_paths(state)` will return a vector (indexed by source vertex) of
+vectors (indexed by destination vertex) of paths. `enumerate_paths(state, v)`
+will return a vector (indexed by destination vertex) of paths from source `v`
+to all other vertices. In addition, `enumerate_paths(state, v, d)` will return
+a vector representing the path from vertex `v` to vertex `d`.
+"""
+function paths(state::ShortestPathResults, vs::AbstractVector{<:Integer})
+    parents = state.parents
+    T = eltype(parents)
+
+    num_vs = length(vs)
+    all_paths = Vector{Vector{T}}(undef, num_vs)
+    for i = 1:num_vs
+        all_paths[i] = Vector{T}()
+        index = T(vs[i])
+        if parents[index] != 0 || parents[index] == index
+            while parents[index] != 0
+                push!(all_paths[i], index)
+                index = parents[index]
+            end
+            push!(all_paths[i], index)
+            reverse!(all_paths[i])
```

Comment:
Maybe we can use `pushfirst!` here, so we don't have to reverse afterwards. As far as I know, Julia also reserves some memory in front of a vector, so this might not be slower.

Reply:
Sounds good. Can you make the change and I'll merge it in?

Reply:
I made the change.